### DISCUSSION TITLE
Messaging semantic: CHILD-OF between producer and consumer spans

### DISCUSSION THREAD

**Discussion by ppatierno** [open-telemetry/opentelemetry-specification#2695] [Q&A] [answered]
Hi all,
I was taking a look at the "messaging semantic conventions" and corresponding examples.
I was also using the Kafka instrumentation library (through interceptors) with a simple Kafka producer sending a message and a Kafka consumer getting that message: the result was a single trace with Kafka receive span as CHILD-OF the Kafka send span.
At same time, I was also looking at the compatibility section with OpenTracing where CHILD-OF is compared to FOLLOW-FROM relationship [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/opentracing.md#span-references) which is the only place where I can find the difference between them.
As per the above line it says that "Child Of: The parent Span depends on the child Span in some capacity." ... so applying this relationship between producer and consumer in a messaging system doesn't sound too right to me. The receive span doesn't really depend on the producer span, as in messaging system everything is totally asynchronous. The producer send the message and the send result doesn't really depends on the corresponding receive operation (which will happen asynchronously and maybe later in time). It is true for Kafka (as I was using it) but also for other messaging systems.
I was wondering what was the rationale to use this CHILD-OF relationship in the messaging semantic conventions examples and in the Kafka instrumentation (I would expect the same result for other instrumented messaging libraries).
Isn't the FOLLOW-FROM a better choice by using OpenTelemetry "link".
Is there anything I am missing here?

**Comment by joaopgrassi** [comment, marked as answer]
Hi @ppatierno !

The [current semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) for messaging systems use span links for some scenarios, like [batching](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-receiving) where the parent-child relationship doesn't really apply.

Moreover, we have a group working on trying to stabilize and better define such conventions for messaging systems. The current work is being done in this [OTEP](https://github.com/open-telemetry/oteps/pull/192).

The TL;DR is that we are going to heavily rely on span links, as we understand they better translate the async nature of messaging systems. So your conclusion pretty much aligns with our current intentions :). 

The group is always eager to get early feedback on what we are proposing, so it would be great if you share your thoughts with us. It can be either by leaving your remarks on the OTEP above or joining or public meetings. We meet every Thursday at 8:00 PT. You can find the details in the community repo (zoom link, meeting notes, etc) https://github.com/open-telemetry/community#specification-sigs.

**Comment by ppatierno** [comment]
Hey @joaopgrassi !

Thank you very much for your quick reply! I will read the OTEP work for sure and come back to you soon or leave my comments there.
Happy to know that it's going to be aligned with my conclusion :-)

**Comment by ppatierno** [comment]
@joaopgrassi just for letting you know that I found really interesting reading the OTEP work and left a few comments.

**Reply by joaopgrassi** [reply]
No problems and thanks for the comments in the OTEP!. We would appreciate if you could join our meetings and share your knowledge with us. Regardless, if my comment here helped clearing the initial question of CHILD-OF, pick any comment that helped you and mark it as an answer so this is ✅ :)

**Reply by ppatierno** [reply]
Yeah I will try to join on Thursday, quite often there are conflicts. Anyway I marked your first comment as answer.
